File: business/src/test/java/com/radioteria/test/business/services/userService/UserRegistrationTest.java
package com.radioteria.test.business.services.userService;

import org.junit.Test;

import static org.mockito.Mockito.*;
import static org.junit.Assert.*;

public class UserRegistrationTest extends AbstractUserServiceTest {

    @Test
    public void testNewUserRegistration() {

//        when(userDao.isEmailAvailable(anyString())).thenReturn(true);
//
//        userService.register("foo@bar.com", "baz", "Foo Bar");
//
//        User user = captureUserThatBeenKept();
//
//        verifyCapturedUser(user);
//        verifyThatEventFiredWithUser(user);

    }

    @Test
    public void testUserRegistrationWhenUserExists() {

//        when(userDao.isEmailAvailable(anyString())).thenReturn(false);
//
//        userService.register("foo@bar.com", "baz", "Foo Bar");

    }

//    private User captureUserThatBeenKept() {
//
//        ArgumentCaptor<User> argumentCaptor = ArgumentCaptor.forClass(User.class);
//
//        verify(userDao, times(1)).persist(argumentCaptor.capture());
//
//        return argumentCaptor.getValue();
//
//    }

//    private void verifyCapturedUser(User user) {
//
//        assertEquals("foo@bar.com", user.getEmail());
//        assertEquals("Foo Bar", user.getName());
//        assertEquals(UserState.INACTIVE, user.getState());
//        assertTrue(userService.isPasswordMatch("baz", user));
//
//    }

//    private void verifyThatEventFiredWithUser(User user) {
//
//        ArgumentCaptor<UserRegisteredEvent> argumentCaptor = ArgumentCaptor.forClass(UserRegisteredEvent.class);
//
//        verify(eventPublisher).publishEvent(argumentCaptor.capture());
//
//        UserRegisteredEvent capturedEvent = argumentCaptor.getValue();
//
//        assertSame(user, capturedEvent.getUser());
//
//    }

}

File: src/test/kotlin/com/radioteria/test/services/user/UserRegistrationTest.kt
package com.radioteria.test.services.user

import com.radioteria.services.user.events.UserRegisteredEvent
import com.radioteria.services.user.exceptions.UserServiceException
import com.radioteria.services.util.TestEventPublisher
import org.junit.Test

import org.springframework.transaction.annotation.Transactional
import kotlin.test.assertTrue

open class UserRegistrationTest : BaseUserServiceTest() {

    @Test
    @Transactional
    open fun registerUser() {
        userService.register("foo@bar.com", "baz", "Foo Bar")

        assertTrue { userService.exists("foo@bar.com") }

        assertTrue { (eventPublisher as TestEventPublisher)
                .isEventPublished(UserRegisteredEvent::class.java) }
    }

    @Test(expected = UserServiceException::class)
    @Transactional
    open fun testUserRegistrationWhenUserExists() {
        userService.register("foo@bar.com", "baz", "Foo Bar")
        userService.register("foo@bar.com", "baz", "Foo Bar")
    }

}
